# Import Libraries

In [ ]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


# Read Dataset

In [ ]:
df = pd.read_csv('Cosmetics.csv')
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Trans.   Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  \
0          1    0      1            1        1          1                0   
1          2    0      0            1        0          1                0   
2          3    0      1            0        0          1                1   
3          4    0      0            1        1          1                0   
4          5    0      1            0        0          1                0   
..       ...  ...    ...          ...      ...        ...              ...   
995      996    0      0            0        0          0                0   
996      997    0      0            0        0          0                0   
997      998    0      1            1        1          1                0   
998      999    1      1            0        0          1                0   
999     1000    0      0            0        0          1                0   

     Bronzer  Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  \
0          1          1        1           0           0          0         0   
1          1          1        0           0           1          1         0   
2          1          1        1           1           1          1         1   
3          1          0        0           0           1          0         0   
4          1          1        1           1           0          1         1   
..       ...        ...      ...         ...         ...        ...       ...   
995        0          0        0           0           0          0         0   
996        0          0        0           0           1          1         0   
997        0          1        0           0           1          0         1   
998        0          0        1           1           0          0         1   
999        1          1        0           1           0          0         0   

     Eyeliner  
0           1  
1           0  
2           0  
3           1  
4           0  
..        ...  
995         0  
996         0  
997         1  
998         1  
999         1  

[1000 rows x 15 columns]

# Pre-process

In [ ]:
df = df.astype(bool)
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Trans.     Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  \
0       True  False   True         True     True       True            False   
1       True  False  False         True    False       True            False   
2       True  False   True        False    False       True             True   
3       True  False  False         True     True       True            False   
4       True  False   True        False    False       True            False   
..       ...    ...    ...          ...      ...        ...              ...   
995     True  False  False        False    False      False            False   
996     True  False  False        False    False      False            False   
997     True  False   True         True     True       True            False   
998     True   True   True        False    False       True            False   
999     True  False  False        False    False       True            False   

     Bronzer  Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  \
0       True       True     True       False       False      False     False   
1       True       True    False       False        True       True     False   
2       True       True     True        True        True       True      True   
3       True      False    False       False        True      False     False   
4       True       True     True        True       False       True      True   
..       ...        ...      ...         ...         ...        ...       ...   
995    False      False    False       False       False      False     False   
996    False      False    False       False        True       True     False   
997    False       True    False       False        True      False      True   
998    False      False     True        True       False      False      True   
999     True       True    False        True       False      False     False   

     Eyeliner  
0        True  
1       False  
2       False  
3        True  
4       False  
..        ...  
995     False  
996     False  
997      True  
998      True  
999      True  

[1000 rows x 15 columns]

In [ ]:
df.drop('Trans. ', axis=1, inplace=True)
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  Bronzer  \
0    False   True         True     True       True            False     True   
1    False  False         True    False       True            False     True   
2    False   True        False    False       True             True     True   
3    False  False         True     True       True            False     True   
4    False   True        False    False       True            False     True   
..     ...    ...          ...      ...        ...              ...      ...   
995  False  False        False    False      False            False    False   
996  False  False        False    False      False            False    False   
997  False   True         True     True       True            False    False   
998   True   True        False    False       True            False    False   
999  False  False        False    False       True            False     True   

     Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  Eyeliner  
0         True     True       False       False      False     False      True  
1         True    False       False        True       True     False     False  
2         True     True        True        True       True      True     False  
3        False    False       False        True      False     False      True  
4         True     True        True       False       True      True     False  
..         ...      ...         ...         ...        ...       ...       ...  
995      False    False       False       False      False     False     False  
996      False    False       False        True       True     False     False  
997       True    False       False        True      False      True      True  
998      False     True        True       False      False      True      True  
999       True    False        True       False      False     False      True  

[1000 rows x 14 columns]

# Generate all the frequent itemsets with support = 20%

In [ ]:
frequent_itemsets = apriori(df, min_support=0.2, use_colnames=True)
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                  itemsets
0     0.363                   (Blush)
1     0.280             (Nail Polish)
2     0.442               (Concealer)
3     0.279                 (Bronzer)
4     0.234               (Lip liner)
5     0.357                 (Mascara)
6     0.381              (Eye shadow)
7     0.536              (Foundation)
8     0.490               (Lip Gloss)
9     0.322                (Lipstick)
10    0.457                (Eyeliner)
11    0.220        (Blush, Concealer)
12    0.204      (Mascara, Concealer)
13    0.201   (Eye shadow, Concealer)
14    0.231   (Concealer, Foundation)
15    0.218    (Lip Gloss, Concealer)
16    0.297     (Concealer, Eyeliner)
17    0.321     (Mascara, Eye shadow)
18    0.211  (Eye shadow, Foundation)
19    0.201   (Eye shadow, Lip Gloss)
20    0.356   (Lip Gloss, Foundation)
21    0.238    (Eyeliner, Foundation)
22    0.227     (Lip Gloss, Eyeliner)

# What is the total number of frequent itemsets?

In [ ]:
len(frequent_itemsets)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


23

# Show all the maximal frequent itemsets.

In [ ]:
frequent_itemsets['Item Set Site'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets.sort_values('support', ascending=False)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                  itemsets  Item Set Site
7     0.536              (Foundation)              1
8     0.490               (Lip Gloss)              1
10    0.457                (Eyeliner)              1
2     0.442               (Concealer)              1
6     0.381              (Eye shadow)              1
0     0.363                   (Blush)              1
5     0.357                 (Mascara)              1
20    0.356   (Lip Gloss, Foundation)              2
9     0.322                (Lipstick)              1
17    0.321     (Mascara, Eye shadow)              2
16    0.297     (Concealer, Eyeliner)              2
1     0.280             (Nail Polish)              1
3     0.279                 (Bronzer)              1
21    0.238    (Eyeliner, Foundation)              2
4     0.234               (Lip liner)              1
14    0.231   (Concealer, Foundation)              2
22    0.227     (Lip Gloss, Eyeliner)              2
11    0.220        (Blush, Concealer)              2
15    0.218    (Lip Gloss, Concealer)              2
18    0.211  (Eye shadow, Foundation)              2
12    0.204      (Mascara, Concealer)              2
19    0.201   (Eye shadow, Lip Gloss)              2
13    0.201   (Eye shadow, Concealer)              2

In [ ]:
frequent_itemsets.drop('Item Set Site', axis=1, inplace=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


# Generate all the association rules with minimum support = 20% and minimum confidence = 70%

In [ ]:
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.7)
rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents   consequents  antecedent support  consequent support  \
0     (Mascara)  (Eye shadow)               0.357               0.381   
1  (Eye shadow)     (Mascara)               0.381               0.357   
2   (Lip Gloss)  (Foundation)               0.490               0.536   

   support  confidence      lift  leverage  conviction  zhangs_metric  
0    0.321    0.899160  2.359999  0.184983    6.138417       0.896222  
1    0.321    0.842520  2.359999  0.184983    4.083050       0.930971  
2    0.356    0.726531  1.355468  0.093360    1.696716       0.514210

# Show rule with highest confidence

In [ ]:
rules.sort_values(by='confidence', ascending=False)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents   consequents  antecedent support  consequent support  \
0     (Mascara)  (Eye shadow)               0.357               0.381   
1  (Eye shadow)     (Mascara)               0.381               0.357   
2   (Lip Gloss)  (Foundation)               0.490               0.536   

   support  confidence      lift  leverage  conviction  zhangs_metric  
0    0.321    0.899160  2.359999  0.184983    6.138417       0.896222  
1    0.321    0.842520  2.359999  0.184983    4.083050       0.930971  
2    0.356    0.726531  1.355468  0.093360    1.696716       0.514210

- Explain the meaning of this rule
 - เมื่อมีการซื้อ Mascara มีความเป็นไปได้ถึง 89.92% ที่ลูกค้าจะซื้อ Eye shadow เพิ่มด้วย ซึ่งบ่งชี้ว่าลูกค้ามักจะเลือกซื้อสินค้าทั้งสองรายการนี้พร้อมกัน
- Explain how support, confidence and lift are calculated?
 - Support: หมายถึงสัดส่วนของธุรกรรมทั้งหมดที่มีสินค้าทั้งสองรายการอยู่ร่วมกัน ยกตัวอย่างเช่น Support 32.1% หมายความว่า 32.1% ของธุรกรรมทั้งหมดประกอบด้วยทั้ง Mascara และ Eye shadow
 - Confidence: คือระดับความเชื่อมั่นของกฎ โดยคำนวณจากจำนวนธุรกรรมที่มีสินค้าทั้งสองรายการร่วมกัน หารด้วยจำนวนธุรกรรมที่มี Mascara
 - Lift: เป็นการประเมินว่าความน่าจะเป็นที่สินค้าสองรายการจะถูกซื้อพร้อมกันสูงเพียงใดเมื่อเทียบกับการซื้อแยกกันอย่างอิสระ
- Show example of using these rules to increase sales.
 - สามารถใช้กฎนี้เพื่อกระตุ้นยอดขายได้โดยการ: จัดโปรโมชั่นหรือจัดวางสินค้าทั้งสองไว้ใกล้กันในร้าน มอบส่วนลดสำหรับการซื้อ Mascara และ Eye shadow พร้อมกัน